tests/importer/importer.cpp

```diff
@@ -0,0 +1,762 @@
+    // it have the same end result.
+    auto validator = libcellml::Validator::create();
+    auto model = libcellml::Model::create("model_from_imports");
+    auto c = libcellml::Component::create("c");
+    auto importer = libcellml::Importer::create();
+
+    auto imp = libcellml::ImportSource::create();
+    imp->setUrl("units_in_cn.cellml");
+
+    c->setImportReference("myComponent");
+    c->setImportSource(imp);
+    model->addComponent(c);
+
+    EXPECT_TRUE(model->hasUnresolvedImports());
+
+    importer->resolveImports(model, resourcePath());
+    EXPECT_FALSE(model->hasUnresolvedImports());
+    model = importer->flatten(model);
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->errorCount());
+}
+
+TEST(Model, importingComponentWithCnUnitsThatAreAlreadyDefinedInImportingModel)
+{
+    const std::string e =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"myModel\">\n"
+        "  <units name=\"myUnitsThatIUse\">\n"
+        "    <unit units=\"metre\"/>\n"
+        "  </units>\n"
+        "  <units name=\"myUnitsThatIUse_1\">\n"
+        "    <unit units=\"second\"/>\n"
+        "  </units>\n"
+        "  <component name=\"c\">\n"
+        "    <variable name=\"a\" units=\"second\"/>\n"
+        "    <math xmlns=\"http://www.w3.org/1998/Math/MathML\" xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\">\n"
+        "      <apply>\n"
+        "        <eq/>\n"
+        "        <ci>a</ci>\n"
+        "        <cn cellml:units=\"myUnitsThatIUse_1\">1</cn>\n"
+        "      </apply>\n"
+        "    </math>\n"
+        "  </component>\n"
+        "</model>\n";
+
+    const std::string in =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"myModel\">\n"
+        "  <units name=\"myUnitsThatIUse\">\n"
+        "    <unit units=\"second\"/>\n"
+        "  </units>\n"
+        "   <component name=\"myComponent\">\n"
+        "       <variable name=\"a\" units=\"second\"/>\n"
+        "       <math xmlns=\"http://www.w3.org/1998/Math/MathML\" xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\">\n"
+        "           <apply>"
+        "             <eq/>\n"
+        "             <ci>a</ci>\n"
+        "             <cn cellml:units=\"myUnitsThatIUse\">1</cn>\n"
+        "           </apply>\n"
+        "       </math>\n"
+        "   </component>\n"
+        "</model>";
+
+    // Create the model by parsing the string above.
+    auto parser = libcellml::Parser::create();
+    auto importedModel = parser->parseModel(in);
+    auto importer = libcellml::Importer::create();
+    auto validator = libcellml::Validator::create();
+
+    // No problems with the imported model.
+    validator->validateModel(importedModel);
+    EXPECT_EQ(size_t(0), validator->errorCount());
+
+    // The model myModel already has myUnitsThatIUse defined.
```

Comment:
?? After line 223, the model looks as follows:
```
<?xml version="1.0" encoding="UTF-8"?>
<model xmlns="http://www.cellml.org/cellml/2.0#" name="myModel">
  <units name="myUnitsThatIUse">
    <unit units="metre"/>
  </units>
</model>
```

So, `myUnitsThatIUse` is not already defined... at this stage. It will be after line 236 when flattening the model. So, may want to update the comment and maybe add another before flattening the model?

Reply:
89b58a38 It's the model name that's wrong in the comment, should be `importedModel`.

Reply:
89b58a3